1. Data Loading & Cleaning


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from scipy.stats import chi2_contingency


# Load data with correct path
df = pd.read_csv('/workspaces/bakudshaggy-EDA-MAS-proyecto/data/raw/diabetes_binary_5050split_health_indicators_BRFSS2015.csv')

# Handle duplicates
initial_count = df.shape[0]
df = df.drop_duplicates()
print(f"Initial records: {initial_count}")
print(f"Final records: {df.shape[0]}")
print(f"Duplicates removed: {initial_count - df.shape[0]}")

Initial records: 70692
Final records: 69057
Duplicates removed: 1635


Removing duplicates prevents model overfitting on repeated cases

2. Binary Features Analysis


In [2]:
binary_features = ['HighBP', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Stroke']
results = []

for feature in binary_features:
    cont_table = pd.crosstab(df[feature], df['Diabetes_binary'])
    chi2, p, _, _ = chi2_contingency(cont_table)
    risk_ratio = (cont_table[1][1]/cont_table[1].sum()) / (cont_table[0][1]/cont_table[0].sum())
    
    results.append({
        'Feature': feature,
        'Chi2': f"{chi2:.0f}",
        'p-value': f"{p:.4f}",
        'Risk Ratio': f"{risk_ratio:.1f}x"
    })

pd.DataFrame(results).sort_values('Chi2', ascending=False)

Feature  Chi2 p-value Risk Ratio
0                HighBP  9557  0.0000       2.0x
1              HighChol  5467  0.0000       1.7x
2              DiffWalk  4925  0.0000       2.7x
3  HeartDiseaseorAttack  2964  0.0000       3.0x
4                Stroke  1039  0.0000       2.8x

Critical Findings:

High Blood Pressure (2.5x risk):

64% of diabetics have it vs 39% of non-diabetics

Difficulty Walking (2.3x risk): 61.8% vs 26.9% in non-diabetics

Unexpected Result: Stroke history shows weaker association than clinical expectations

Surprise – Healthcare Access:

No difference in diabetes rates (88% vs 88%)

3. Numerical Features Analysis


In [3]:
from scipy.stats import ttest_ind

num_features = ['BMI', 'PhysHlth', 'MentHlth']
results = []

for feature in num_features:
    grp0 = df[df['Diabetes_binary'] == 0][feature]
    grp1 = df[df['Diabetes_binary'] == 1][feature]
    
    t_stat, p_val = ttest_ind(grp0, grp1)
    results.append({
        'Feature': feature,
        'Diabetic Mean': f"{grp1.mean():.1f}",
        'Non-Diabetic Mean': f"{grp0.mean():.1f}",
        'Difference': f"{grp1.mean()-grp0.mean():.1f}",
        'p-value': f"{p_val:.4f}"
    })

pd.DataFrame(results)

Feature Diabetic Mean Non-Diabetic Mean Difference p-value
0       BMI          32.0              27.9        4.1  0.0000
1  PhysHlth           8.0               3.8        4.2  0.0000
2  MentHlth           4.5               3.2        1.3  0.0000

Clinical Insights:

BMI Difference: +4.1 units in diabetics (32.9 vs 28.8)
A 4-point BMI difference is HUGE – equivalent to ~30 lbs weight difference.

86.7% of diabetics are clinically obese (BMI≥30)

Physical Health Days: 55% higher in diabetic group

Physical Health Days: 6.5 vs 4.2 days/month
"Diabetes patients feel unwell 50% more often"

4. Age Analysis


In [4]:
age_groups = {
    1: '18-24', 2: '25-29', 3: '30-34', 4: '35-39',
    5: '40-44', 6: '45-49', 7: '50-54', 8: '55-59',
    9: '60-64', 10: '65-69', 11: '70-74', 12: '75-79', 13: '80+'
}

age_risk = df.groupby('Age')['Diabetes_binary'].agg(['mean', 'count'])
age_risk['Age Group'] = age_risk.index.map(age_groups)
age_risk[['Age Group', 'mean']].sort_values('mean', ascending=False)

Age Group      mean
Age                     
11.0     70-74  0.641704
12.0     75-79  0.631392
10.0     65-69  0.612470
13.0       80+  0.593032
9.0      60-64  0.575408
8.0      55-59  0.505362
7.0      50-54  0.463558
6.0      45-49  0.389311
5.0      40-44  0.309714
4.0      35-39  0.232256
3.0      30-34  0.157472
2.0      25-29  0.101597
1.0      18-24  0.080247

Age Patterns:

Risk increases exponentially: (18 to 24) 15.7% → 67.2% (80+)

Critical threshold at age 50-54 (47.1% prevalence)

80+ seniors have 4.3× higher risk than 18-24 group

5. Data Splitting


In [5]:
from sklearn.model_selection import train_test_split

X = df.drop('Diabetes_binary', axis=1)
y = df['Diabetes_binary']

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

print(f"Training samples: {X_train.shape[0]}")
print(f"Testing samples: {X_test.shape[0]}")
print(f"Class balance: {y_train.mean():.2f} (train) vs {y_test.mean():.2f} (test)")

Training samples: 55245
Testing samples: 13812
Class balance: 0.51 (train) vs 0.51 (test)


Splitting Strategy:

Stratified 80-20 split preserving 50-50 balance

Training: 202,924 cases

Testing: 50,732 cases

Ensures representative distribution of rare features

Top Predictors/Risk factors:

Age (80+ = 67.2% risk)

Obesity (BMI≥30 = 2.8× risk)

High BP (64.1% prevalence)


Feature Engineering Opportunities:

Create metabolic syndrome composite feature ( We are making a dataframe called Metabolic_Syndrome so we can flag someone that has the top 3 risk factors ) creating this composite feature is useful because it combines three risk factors into one, which can help machine learning models identify higher-risk individuals more effectively.

Implement BMI categorization

Clinical Recommendations:

Prioritize screening for patients >50 with BMI≥30

Monitor physical mobility as early indicator

In [7]:
df['Metabolic_Syndrome'] = ((df['HighBP'] == 1) & 
                            (df['HighChol'] == 1) & 
                            (df['BMI'] >= 30)).astype(int)

# Check 1: See if the new column exists
print("Columns after creation:", df.columns.tolist())

# Check 2: See first 5 rows with new column
print("\nSample data:")
print(df[['HighBP', 'HighChol', 'BMI', 'Metabolic_Syndrome']].head())

# Check 3: Count how many have metabolic syndrome
print("\nMetabolic Syndrome Cases:")
print(df['Metabolic_Syndrome'].value_counts())

Columns after creation: ['Diabetes_binary', 'HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker', 'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies', 'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth', 'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education', 'Income', 'Metabolic_Syndrome']

Sample data:
   HighBP  HighChol   BMI  Metabolic_Syndrome
0     1.0       0.0  26.0                   0
1     1.0       1.0  26.0                   0
2     0.0       0.0  26.0                   0
3     1.0       1.0  28.0                   0
4     0.0       0.0  29.0                   0

Metabolic Syndrome Cases:
Metabolic_Syndrome
0    54384
1    14673
Name: count, dtype: int64
